README.md
# rbhex-core

[![Gem Version](https://badge.fury.io/rb/rbhex-core.svg)](http://badge.fury.io/rb/rbhex-core)

Contains core widgets and infrastructure of rbhex ncurses toolkit. rbhex helps to easily build
ncurses applications for text terminals.

rbhex-core contains basic widgets for constructing command line applications with ncurses.  These include:

* field

* buttons - checkbox, radio, toggle

* list

* textview

* dialogs and popup

* table

* menu

* tabbedpane

* tree

* application header

* status line

## Install

`gem install rbhex-core`

## Documentation

In short, there isn't any - nor are there any plans to create any. The `/examples` directory
contains many examples and should provide a nice start for you. Otherwise, the [tutorial for
rbcurse](https://github.com/rkumar/rbcurse-tutorial) is still mostly valid for the majority of this
codebase. 

## rbcurse

This project is a continuation/fork of [rkumar](https://github.com/rkumar)'s [rbcurse](https://github.com/rkumar/rbcurse-core). 
rbcurse is no longer maintained as of March 2014.

## Contributing

1. Fork it ( https://github.com/ChrisArcand/rbhex-core/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

[1]: https://github.com/ChrisArcand/rbhex-core


CHANGELOG
**2014-09-29**
## rbhex-core v1.0.0
  * Redistributed as rbhex
  * Fixed pathing in examples so you can run them standalone
  * Some non-code restructuring of gemspec, etc

**2013-04-01**
## 0.0.14 rbcurse-core
   * added module Rbhex to various modules that were coming as top layer
     Things could break !!!
   * minor fix in rlist list()

**2013-03-29**
## 0.0.13 rbcurse-core
   * textpad - support cases where dimension changes, or a row is updated

**2013-03-25**
## 0.0.12 rbcurse-core
   * Minor tweaks and fixes to statusline (tested in ribhu gem)

**2013-03-22**
## 0.0.11 rbcurse-core
   * Added textpad (tested in cygnus gem)

**2013-03-22**
## 0.0.10 rbcurse-core
   * Added method_missing to chunkline, required for ribhu gem

**2013-03-21**
## 0.0.9 rbcurse-core
   * Made several fixed so it works on 1.8.7 (thanks hramrach)

**2012-01-  **
## 0.0.4 rbcurse-core
   * added RESIZE event to form, so resizing app components is possible
	 when window/terminal is resized
	 * changes to ncurses setting as a result of above (wtimeout commented off)
	 * Moving help system to Form
	 * minor addition of define_prefix_command to Form - experimental

**2012-01-04**
## 0.0.3 rbcurse-core
   * ActionManager added to widgets so actions can be
used from multiple places
   * changes in ncurses setting cbreak and halfdelay etc
	 since C-c was crashing program out
	 * Added history to field

**2011-12-13**
## 0.0.2 rbcurse-core
Mostly refactoring of multi-row components

## 0.0.1 rbcurse-core
Added help system, and printing of key bindings

**2011-12-05**
## 0.0.0 rbcurse-core

1. Added a help system - on F1.
   '?' reveals key-bindings for form and current widget.

2. bind_key includes string parameter which documents binding

3. KEY_RETURN and KEY_ENTER fixed (values switched)

4. textview can extend MultiBuffer (simple multibuffering)

5. Moving map_keys prior to super in constructor of widgets
   so app can override keys.

6. textview: set_content second param is a Hash.

7. App.rb: shortcuts and stack flow removed and replaced
   with widgetshortcuts.rb.
	 Now there should be an outer flow or stack. Earlier,
	 there could be two stacks with no outer stack or flow.

	 New implementation allows height_pc and width_pc.
	 Widgets can also have height_pc and width_pc if placed in stacks or flows
	 in an App. Please check samples to see.

8. Window has close event, to do any tidying up, releasing etc
   Also a confirm_close option.

**2011-11-17**
## 1.5.0

### split rbcurse into core, extras and experimental
    Now rbcurse is a meta-gem which includes the other three.

### major rehaul of directory structures
    ilb/core/widgets, lib/core/include, lib/core/system etc

### class name changes

    * rbasiclistbox -> rlist, shortcut: basic_list -> listbox
		* rlistbox              , shortcut: list_box   -> edit_list
    * rtable                , shortcut: table      -> edit_table
    * rtabularwidget        , shortcut: tabular_widget -> table (alias)


**2011-11-06**
## 1.4.1

### More cleanup of widget and form and window

### Rewrite of tabbedpane and messagebox but not replaced old classes

### Textview and Window and Statusline now can print colored text
    Colored interface not yet final.

### Some experimental work that will trickle into core classes in 1.5.0

### Some standardization of core classes, more to follow.

**2011-10-06**
## 1.3.1 or 1.4.0

### Major cleanup of widget class
  Scrollpane support (buffering stuff all removed)

### TabbedPane
  Corrected, major changes

### dsl_accessor and dsl_property
  Now return self so we can chain.

### return values of some keys (Class Form)
  Earlier not returning anything meaningful or returning last_key
  Now return NO_NEXT_FIELD or NO_PREVIOUS_FIELD

### j and k keys used for down and up where possible
   e.g. button and its children including checkbox

### Miscellaneous

  * Dynamic menuitems in menu (see menu1.tb)
  * rework on menu and menubar, and keys using them
  * FieldVetoException (see test2.rb)
  * valid_range for Field (see test2.rb)
  * fixed field's type method, earlier non-functional
  * array and 4 integer constructor for Window
  * StatusWindow class - 2 line window for alert, confirm and statuses as an
    alternative to dialogs.
  * status_dialog returning a handle so we can keep updating status then
  close.
  * many more minor adjustments, tweaks, issues

**2011-09-26**
## 1.3.0

### Moving to ffi-ncurses
   Essentially code changes for ffi-ncurses. ncurses gem no longer supported
   as it was painful to install. Required ffi-ncurses >= 0.4.0

   Some bugfixes in demos and samples due to changes in 1.2.0, or ruby 1.9,  which went
   unnoticed.

### Deprecations
   Some classes being deprecated:
   - SplitPane (use vimsplit instead)
   - ScrollPane (no substitute) Code supporting scrollpane will be removed
   by 1.3.1 or 1.3.2  and this will not work.

### Tabbedpane
   Not working correctly since moving to ffi-ncurses. Will be completely
   rewritten by 1.3.2.

**2010-09-11**
## 1.2.0

### Additions

 - App class that allows easy application development (like Shoes)
 - MultiSplit class: Multicolumn widget (like SplitPane but allows many splits).
   aka Column Browse pattern (like OSX Finder). See examples/testmultisplit.rb
 - Added Progress Bar class

 - Added methods to Menu (rmenu.rb), so easy creation through DSL. See
 examples/menu1.rb

 - Added VimSplit - can have multiple splits stacked and flowed
 - Added Tree     - tree control
 - Added Scrollbar class
 - Added Tabular - quick readonly tabular data generator that can be used with
 listboxes or textviews.
 - Added BasicList - a stripped down, readonly version of Listbox.
 -

### Changes

 - Field class now passes InputDataEvent when firing event, not self.
   Earlier, user trapping :CHANGE would not know what happened (delete,
   insert, etc) and what char was added, deleted.
 - List and ListData and ListSelection listener

examples/dbdemo.rb
require 'rbhex/core/util/app'
require 'sqlite3'
#require 'rbhex/experimental/resultsettextview.rb'
#require 'rbhex/experimental/widgets/undomanager'

# @return array of table names from selected db file
def get_table_names
  raise "No database file selected." unless $current_db

  $tables = get_data "select name from sqlite_master"
  $tables.collect!{|x| x[0] }  ## 1.9 hack, but will it run on 1.8 ??
  $tables
end
def get_column_names tbname
  get_metadata tbname
end
def connect dbname
  $log.debug "XXX:  CONNECT got #{dbname} "
  $current_db = dbname
  $db = SQLite3::Database.new(dbname) if dbname

  return $db
end
def get_data sql
  $log.debug "SQL: #{sql} "
  $columns, *rows = $db.execute2(sql)
  $log.debug "XXX COLUMNS #{sql}  "
  content = rows
  return nil if content.nil? or content[0].nil?
  $datatypes = content[0].types #if @datatypes.nil?
  return content
end
def get_metadata table
  get_data "select * from #{table} limit 1"
  #$columns.collect!{|x| x[0] }  ## 1.9 hack, but will it run on 1.8 ??
  return $columns
end
#
# creates a popup for selection given the data, and executes given block with
#  following return value.
# @return [String] if mode is :single
# @return [Array] if mode is :multiple
#
def create_popup array, selection_mode=:single,  &blk
  #raise "no block given " unless block_given?
  listconfig = {'bgcolor' => 'blue', 'color' => 'white'}
  listconfig[:selection_mode] = selection_mode
  ix = popuplist array, listconfig
  if ix
    if selection_mode == :single
      value = array[ix]
      blk.call value
    else
      #values = array.select {|v| ix.include? v}
      values = []
      array.each_with_index { |v, i| values << v if ix.include? i }
      blk.call(values)
    end
  end
end

#
# changed order of name and fields, thanks hramrach
def view_data name, fields="*"
  stmt = "select #{fields} from #{name}"
  stmt << $where_string if $where_string
  stmt << $order_string if $order_string
  view_sql stmt
  @form.by_name['tarea'] << stmt if @form # nil when called from menu
end
def view_sql stmt
  begin
  content = get_data stmt
  if content.nil?
  else
    require 'rbhex/core/widgets/tabular'
    t = Tabular.new do |t|
      t.headings = $columns
      t.data=content
    end
    view t.render
  end
  rescue => err
    $log.error err.to_s
    $log.error(err.backtrace.join("\n"))
    textdialog [err.to_s, *err.backtrace], :title => "Exception"
  end
end

App.new do
  #header = app_header "rbhex #{Rbhex::Core::VERSION}", :text_center => "Database Demo", :text_right =>"enabled"
  form = @form
  mylabel = "a field"
  $catch_alt_digits = true # use M-1..9 in textarea
  $current_table = nil
  $current_db = nil # "testd.db"
  connect $current_db if $current_db
  def which_field
    alert "curent field is #{form.get_current_field} "
  end

  def get_commands
    %w{ which_field }
  end
  def help_text
    <<-eos
               DBDEMO HELP

      This is some help text for dbdemo.
      We are testing out this feature.

      Alt-d    -   Select a database
      <Enter>      on a table, view data (q to close window)
      <Space>      on a table, display columns in lower list

                COLUMN LIST KEYS
      <Space>      on a column for multiple select
      <Ctrl-Space> on a column for range select/deselect from previous selection
      <Enter>      on column table to view data for selected columns
             u     unselect all
             a     select all
             *     invert selection
      F4           View data for selected table (or columns if selected)

      q or C-q     Close the data window that comes on Enter or F4

      Alt-x    -   Command mode (<tab> to see commands and select)
      :        -   Command mode
      Alt-z    -   Commands in TextArea

                Sql Entry Area
      C-x e        Edit in $EDITOR or vi
      M-?          To see other key-bindings
      F4           Execute SQL (there should be only one sql).


                Result Set  (this is not present in this demo any longer - moved
                to rbhex-extras)
      ,            Prev row (mnemonic <)
      .            Next row (mnemonic >)
      <            First row
      >            Last row

      F10      -   Quit application



      -----------------------------------------------------------------------
      Hope you enjoyed this help.
    eos
  end
  def ask_databases
      names = Dir.glob("*.{sqlite,db}")
      if names
        ix = popuplist( names )
        if ix
          value = names[ix]
          connect(value);
          @form.by_name["tlist"].list(get_table_names)
        end

      else
        alert "Can't find a .db or .sqlite file"
      end
  end
  @form.help_manager.help_text = help_text()
  # TODO accelerators and
  # getting a handle for later use
  mb = menubar do
    keep_visible true
    #@toggle_key=KEY_F2
    menu "File" do
      item "Open", "O" do
        accelerator "Ctrl-O"
        command do
          alert "HA!! you wanted to open a file?"
        end
      end
      menu "Database" do
        item_list do
          Dir.glob("**/*.{sqlite,db}")
        end
        command do |menuitem, text|
          connect text
          form.by_name["tlist"].list(get_table_names)
        end
      end
      menu "Tables" do
        item_list do
          if $current_db
            get_table_names
          end
        end
        command do |menuitem, text|
          $current_table = text
          #alert(get_column_names(text).join(", "))
          create_popup(get_column_names(text), :multiple) { |value| view_data( text, value.join(",") ) }
        end
      end
      item "New", "N"
      separator
      item "Exit", "x"  do
        command do
          throw(:close)
        end
      end
      item "Cancel Menu" do
        accelerator "Ctrl-g"
      end

    end # menu
    menu "Window" do
      item "Tile", "T"
      menu "Find" do
        item "More", "M"
        $x = item "Less", "L" do
          #accelerator "Ctrl-X"
          command do
            alert "You clickses on Less"
          end
        end
        menu "Size" do
          item "Zoom", "Z"
          item "Maximize", "X"
          item "Minimize", "N"
        end
      end
    end
    menu "Others" do
      require 'rbhex/core/include/appmethods.rb'
      item "Shell Output" do
        command { shell_output }
      end
      item "Suspend" do
        command { suspend }
      end
    end
  end # menubar
  mb.toggle_key = FFI::NCurses::KEY_F2
  mb.color = :white
  mb.bgcolor = :blue
  @form.set_menu_bar mb
  tv = nil
  flow :margin_top => 1 do
    col1w = 20
    stack :width_pc => 20 do
      text = "No tables"
      if !$current_db
        text = "Select DB first. Press Alt-D"
      end
      tlist = basiclist :name => "tlist", :list => [text], :title => "Tables", :height => 10,
        :selected_color => 'cyan', :selected_bgcolor => 'black' , :selected_attr => Ncurses::A_REVERSE,
        :help_text => "Enter to View complete table, Space to select table and view columns",
        :should_show_focus => true,
        :selection_mode => :single
      tlist.bind(:PRESS) do |eve|
        if $current_db
          # get data of table
          view_data eve.text
          #tv.sqlite $current_db, eve.text, "select * from #{eve.text} " # TODO in core
        else
          ask_databases
        end
      end
      #tlist.bind(:ENTER_ROW) do |eve|
        # too much confusion between selected and focussed row
        #$current_table = eve.text if $db
      #end
      clist = basiclist :name => "clist", :list => ["No columns"], :title => "Columns", :height => 14,
        :selection_mode => :multiple,
        :selected_color => 'cyan', :selected_bgcolor => 'black' , :selected_attr => Ncurses::A_REVERSE,
        :help_text => "Enter to View selected fields, Space to select columns, w - where, o-order"
      tlist.bind(:LIST_SELECTION_EVENT) do |eve|
        $selected_table = eve.source[eve.firstrow]
        $current_table = $selected_table
        clist.data = get_column_names $selected_table
      end
      clist.bind(:PRESS) do |eve|
        # get data of table
        if $selected_table
          cols = "*"
          c = clist.get_selected_values
          unless c.empty?
            cols = c.join(",")
          end
          view_data $selected_table, cols
        else
          alert "Select a table first."
        end
      end
      clist.bind_key('w', 'add to where condition') {
        c = clist.current_value
        $where_columns ||= []
        hist = ["#{c} = "]
        w = rb_gets("where "){ |q| q.default = "#{c} = "; q.history = hist }
        $where_columns << w if w
        message "where: #{$where_columns.last}. Press F4 when done"
        $log.debug "XXX: WHERE: #{$where_columns} "
      }
      clist.bind_key('o', 'add to order by') {
        c = clist.current_value
        $order_columns ||= []
        $order_columns << c if c
        message "order (asc): #{$order_columns.last}. Press F4 when done"
        $log.debug "XXX: ORDER: #{$order_columns} "
      }
      clist.bind_key('O', 'add to ordery by desc') {
        c = clist.current_value
        $order_columns ||= []
        $order_columns << " #{c} desc " if c
        message "order: #{$order_columns.last}"
        $log.debug "XXX: ORDER: #{$order_columns}. Press F4 when done"
      }
      @statusline = status_line
      #wg = get_color($datacolor, 'white','green')
      #wb = get_color($datacolor, 'white','blue')
      @statusline.command {
        # trying this out. If you want a persistent message that remains till the next on
        #  then send it in as $status_message
        text = $status_message.value || ""
        if !$current_db
          #"[%-s] %s" % [ "Select a Database", text]
          "[%-s] %s" % [ "#[bg=red,fg=yellow]Select a Database#[end]", text]
          #[ [nil, "%-22s" % Time.now, nil], [$errorcolor, " [Select a Database ]", FFI::NCurses::A_BOLD], [nil, text, nil] ]
        elsif !$current_table
          "[DB: #[fg=white,bg=blue]%-s#[end] | %-s ] %s" % [ $current_db || "None", $current_table || "#[bg=red,fg=yellow]Select a table#[end]", text]
          #[ [nil, "%-22s [DB: %-s | " % [Time.now, $current_db || "None" ],nil], [$errorcolor, " Select a Table ]", FFI::NCurses::A_BOLD], [nil, text, nil] ]
        else
          "DB: #[fg=white,bg=green,bold]%-s#[end] | #[bold]%-s#[end] ] %s" % [ $current_db || "None", $current_table || "----", text]
          #[ [nil, "%-22s [DB: " % Time.now, nil], [wb, " #{$current_db} ", FFI::NCurses::A_BOLD],
          #[wg, $current_table || "----", FFI::NCurses::A_BOLD], [nil, text, nil] ]
        end
      }
      @adock = nil
      keyarray = [
        ["F1" , "Help"], ["F10" , "Exit"],
        ["F2", "Menu"], ["F4", "View"],
        ["M-d", "Datebase"], ["M-t", "Table"],
        ["M-x", "Command"], nil
      ]
      tlist_keyarray = keyarray + [ ["Sp", "Select"], nil, ["Enter","View"] ]

      clist_keyarray = keyarray + [ ["Sp", "Select"], ["C-sp", "Range Sel"],
        ["Enter","View"], ['w', 'where'],
        ["o","order by"], ['O', 'order desc']
      ]
      tarea_keyarray = keyarray + [ ["M-z", "Commands"], nil ]
      #tarea_sub_keyarray = [ ["r", "Run"], ["c", "clear"], ["w","Save"], ["a", "Append next"],
      #["y", "Yank"], ["Y", "yank pop"] ]
      tarea_sub_keyarray = [ ["r", "Run"], ["c", "clear"], ["e", "Edit externally"], ["w","Kill Ring Save (M-w)"], ["a", "Append Next"],
        ["y", "Yank (C-y)"], ["Y", "yank pop (M-y)"],
        ["u", "Undo (C-_)"], ["R", "Redo (C-r)"],
      ]

      gw = get_color($reversecolor, 'green', 'black')
      @adock = dock keyarray, { :row => Ncurses.LINES-2, :footer_color_pair => $datacolor,
        :footer_mnemonic_color_pair => gw }
      @adock.set_key_labels tlist_keyarray, :tables
      @adock.set_key_labels clist_keyarray, :columns
      @adock.set_key_labels tarea_sub_keyarray, :tarea_sub
      @adock.set_key_labels tarea_keyarray, :tarea
      tlist.bind(:ENTER) { @adock.mode :tables }
      clist.bind(:ENTER) { @adock.mode :columns }

      reduce = lambda { |obj|
        obj.height -= 1 if obj.height > 3
      }
      increase = lambda { |obj|
        obj.height += 1 if obj.height + obj.row < Ncurses.LINES-2
      }
      _lower = lambda { |obj|
        obj.row += 1 if obj.height + obj.row < Ncurses.LINES-2
      }
      _raise = lambda { |obj|
        obj.row -= 1 if obj.row > 2
      }
      [clist, tlist].each do |o|
        o.bind_key([?\C-x, ?-]){ |o| reduce.call(o) }
        o.bind_key([?\C-x, ?+]){ |o| increase.call(o) }
        o.bind_key([?\C-x, ?v]){ |o| _lower.call(o) }
        o.bind_key([?\C-x, ?6]){ |o| _raise.call(o) }
      end


      @form.bind_key([?q,?q], 'quit') { throw :close }
      @form.bind_key(?\M-t, 'select table') do
        if $current_db.nil?
          alert "Please select database first"
        else
          create_popup( get_table_names,:single) {|value| $selected_table = $current_table =  value}
        end
      end
      @form.bind_key(?\M-d, 'select database') do
        ask_databases
      end
      @form.bind_key(FFI::NCurses::KEY_F4, 'view data') do
        $where_string = nil
        $order_string = nil
        if $where_columns
          $where_string = " where " + $where_columns.join(" and ")
        end
        if $order_columns
          $order_string = " order by " + $order_columns.join(" , ")
        end
        # mismatch between current and selected table
        if $current_table
          cols = "*"
          c = clist.get_selected_values
          unless c.empty?
            cols = c.join(",")
          end
          view_data $current_table, cols
        else
          alert "Select a table first."
        end
      end
    end # stack
    stack :width_pc => 80 do
      tarea = textarea :name => 'tarea', :height => 5, :title => 'Sql Statement'
      #undom = SimpleUndo.new tarea
      tarea.bind_key(Ncurses::KEY_F4, 'view data') do
        text = tarea.get_text
        if text == ""
          alert "Please enter a query and then hit F4. Or press F4 over column list"
        else
          view_sql tarea.get_text
        end
      end
      tarea.bind(:ENTER) { @adock.mode :tarea }
      tarea.bind_key(?\M-z, 'textarea submenu'){

        hash = { 'c' => lambda{ tarea.remove_all },
          'e' => lambda{ tarea.edit_external },
          'w' => lambda{ tarea.kill_ring_save },
          'a' => lambda{ tarea.append_next_kill },
          'y' => lambda{ tarea.yank },
          'Y' => lambda{ tarea.yank_pop },
          'r' => lambda{ view_sql tarea.get_text },
          'u' => lambda{ tarea.undo },
          'R' => lambda{ tarea.redo },
      }


      @adock.mode :tarea_sub
      @adock.repaint
      keys = @adock.get_current_keys
      while((ch = @window.getchar()) != ?\C-c.getbyte(0) )
        if ch < 33 || ch > 126
          Ncurses.beep
        elsif !keys.include?(ch.chr)
          Ncurses.beep
        else
          hash.fetch(ch.chr).call
          #opt_file ch.chr
          break
        end
      end
      @adock.mode :normal
      } # M-z
      flow do
        #button_row = 17
        button "Save" do
          @cmd_history ||= []
          filename = rb_gets("File to append contents to: ") { |q| q.default = @oldfilename; q.history = @cmd_history }

          if filename
            str = tarea.get_text
            File.open(filename, 'a') {|f| f.write(str) }
            @oldfilename = filename
            @cmd_history << filename unless @cmd_history.include? filename

            message "Appended data to #{filename}"
          else
            message "Aborted operation"
          end
          #hide_bottomline
        end
        button "Read" do
          filter = "*"
          #str = choose filter, :title => "Files", :prompt => "Choose a file: "
          cproc = Proc.new { |str| Dir.glob(str + "*") }
          str = rb_gets "Choose a file: ", :title => "Files", :tab_completion => cproc,
            :help_text => "Press <tab> to complete filenames. C-a, C-e, C-k. Alt-?"
          if str && File.exists?(str)
            begin
              tarea.set_content(str)
              message "Read content from #{str} "
            rescue => err
              print_error_message "No file named: #{str}: #{err.to_s} "
            end
          end
        end
        #ok_button = button( [button_row,30], "OK", {:mnemonic => 'O'}) do
        #end
      end
      blank
      #tv = Rbhex::ResultsetTextView.new @form, :row => 1,  :col => 1, :width => 50, :height => 16
      #tv = resultsettextview :name => 'resultset', :height => 18 , :title => 'DB Browser', :print_footer => true

    end
  end
end # app
